lib/experimental/html/class-wp-html-tag-processor.php

```diff
@@ -948,6 +948,38 @@ public function set_attribute( $name, $value ) {
 			return;
 		}
 
+		/*
+		 * Verify that the attribute name is allowable. In WP_DEBUG
+		 * environments we want to crash to quickly alert developers
+		 * of typos and issues; but in production we don't want to
+		 * interrupt a normal page view, so we'll silently avoid
+		 * updating the attribute in those cases.
+		 *
+		 * Of note, we're disallowing more characters than are strictly
+		 * forbidden in HTML5. This is to prevent additional security
+		 * risks deeper in the WordPress and plugin stack. Specifically
+		 * we reject the less-than (<) and ampersand (&) characters.
+		 *
+		 * The use of a PCRE match allows us to look for specific Unicode
+		 * code points without writing a UTF-8 decoder. Whereas scanning
+		 * for one-byte characters is trivial, scanning for the longer
+		 * byte sequences would be more complicated, and this shouldn't
+		 * be in the hot path for execution so we can compromise on the
+		 * efficiency at this point.
+		 *
+		 * @see https://html.spec.whatwg.org/#attributes-2
+		 */
+		if ( preg_match(
+			'~[ "\'>&</=\x{00}-\x{1F}\x{FDD0}-\x{FDEF}\x{FFFE}\x{FFFF}\x{1FFFE}\x{1FFFF}\x{2FFFE}\x{2FFFF}\x{3FFFE}\x{3FFFF}\x{4FFFE}\x{4FFFF}\x{5FFFE}\x{5FFFF}\x{6FFFE}\x{6FFFF}\x{7FFFE}\x{7FFFF}\x{8FFFE}\x{8FFFF}\x{9FFFE}\x{9FFFF}\x{AFFFE}\x{AFFFF}\x{BFFFE}\x{BFFFF}\x{CFFFE}\x{CFFFF}\x{DFFFE}\x{DFFFF}\x{EFFFE}\x{EFFFF}\x{FFFFE}\x{FFFFF}\x{10FFFE}\x{10FFFF}]~Ssu',
+			$name
+		) ) {
+			if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
+				throw new Exception( 'Invalid attribute name' );
+			}
+
+			return;
```

Comment:
return new WP_Error perhaps?

Reply:
I don't think we should return an object as devs won't check the return value anyway.

The special case in the debug mode looks like a reasonable approach. I'm curious whether there is any standardized way of logging those issues. 

Maybe we could call [`_doing_it_wrong`](https://developer.wordpress.org/reference/functions/_doing_it_wrong/). We use that in several places when registering blocks or similar stuff for the block editor, for example:

https://github.com/WordPress/wordpress-develop/blob/2b1febd20d77898eb81439a688ea5597da00172a/src/wp-includes/class-wp-block-type-registry.php#L55L62

Reply:
this is the big point about this patch that I'm unsure of. my only goal is to quickly bail if someone submits the wrong attribute, as I don't think it's likely people will send user-input attribute names here. if they _do_ do that then I don't want it to crash the full render.

I'll ask around and see if anyone else has experience with this kind of failure.

Reply:
> my only goal is to quickly bail if someone submits the wrong attribute

Yes, think this is the best approach here. If a plugin allows user input for attributes names (very unlikely as usually these are not random), it will have to make sure the input is valid, or the attribute will be skipped.

Reply:
> I'm curious whether there is any standardized way of logging those issues.

Think there isn't one, unfortunately. Has been discussed few times afaik but a decision was not reached. Thinking it is time for this to be added to core. Perhaps another constant: `WP_DEV_MODE` or similar, then be much bolder about throwing errors and exceptions (with backtrace?) and writing in logs.

Using `WP_DEBUG` seems proper here imho. `doing_it_wrong` doesn't seem as good because it is targeted more at developers that try to use some function/method improperly (when there are better ways).

Reply:
@azaozz, great feedback! Thank you so much for clarifying where given options fit best.